src/models/stores/stores.ts

```diff
@@ -35,142 +40,174 @@ export interface ICreateStores extends Partial<IStores> {
+    this.db = params?.db || new DB();
+    this.documents = params?.documents || createDocumentsModelWithRequiredDocuments(requiredDocumentTypes);
+    this.networkDocuments = params?.networkDocuments || DocumentsModel.create({});
+    this.unit = params?.unit || UnitModel.create({code: "NULL", title: "Null Unit"});
+    const demoName = params?.demoName || this.appConfig.appName;
+    this.demo = params?.demo || DemoModel.create({name: demoName, class: {id: "0", name: "Null Class"}});
+    this.showDemoCreator = params?.showDemoCreator || false;
+    this.supports = params?.supports || SupportsModel.create({});
+    this.clipboard = ClipboardModel.create();
+    this.selection = SelectionStoreModel.create();
+    this.serialDevice = new SerialDevice();
+    this.ui.setProblemPath(this.problemPath);
+    this.userContextProvider = new UserContextProvider(this);
+  }
 
-export function getTabsToDisplay(stores: IBaseStores) {
-  const { appConfig: { navTabs: navTabSpecs },
-    teacherGuide,
-    user: { isTeacher }
-  } = stores;
+  get tabsToDisplay() {
+    const { appConfig: { navTabs: navTabSpecs },
+      teacherGuide,
+      user: { isTeacher }
+    } = this;
 
-  return isTeacher
-    ? navTabSpecs.tabSpecs.filter(t => (t.tab !== "teacher-guide") || teacherGuide)
-    : navTabSpecs.tabSpecs.filter(t => !t.teacherOnly);
-}
+    return isTeacher
+      ? navTabSpecs.tabSpecs.filter(t => (t.tab !== "teacher-guide") || teacherGuide)
+      : navTabSpecs.tabSpecs.filter(t => !t.teacherOnly);
+  }
 
-export const setUnitAndProblem = async (stores: IStores, unitId: string | undefined, problemOrdinal?: string) => {
-  let unitJson = await getUnitJson(unitId, stores.appConfig);
-  if (unitJson.status === 404) {
-    unitJson = await getUnitJson(stores.appConfig.defaultUnit, stores.appConfig);
+  get problemPath() {
+    return `${this.unit.code}/${this.investigation.ordinal}/${this.problem.ordinal}`;
   }
 
-  // read the unit content, but don't instantiate section contents (DocumentModels) yet
-  const unit = createUnitWithoutContent(unitJson);
+  get problemOrdinal() {
+    const { investigation, problem } = this;
+    return `${investigation.ordinal}.${problem.ordinal}`;
+  }
 
-  const _problemOrdinal = problemOrdinal || stores.appConfig.defaultProblemOrdinal;
-  const { investigation: _investigation, problem: _problem } = unit.getProblem(_problemOrdinal);
+  get isShowingTeacherContent() {
+    const { ui: { showTeacherContent }, user: { isTeacher } } = this;
+    return isTeacher && showTeacherContent;
+  }
 
-  stores.appConfig.setConfigs([unit.config || {}, _investigation?.config || {}, _problem?.config || {}]);
+  // If we need to batch up the changes here we could try changing this to a
+  // MobX flow. However we need to experiment with the interaction of MobX flows
+  // with MST flows.
+  async setUnitAndProblem(unitId: string | undefined, problemOrdinal?: string) {
```

Comment:
Shouldn't this be declared an `action` up above?

Reply:
Because this is async it is complicated. The best solution would be to make it a `flow` but typing flows is annoying. Because I switched to `makeAutoObservable` I'm not sure how it is handled now, but I'm pretty sure it won't really be in an action after the first await call which is at the top of the function. So all of the places where properties are set are now wrapped in `runInAction`.